part1/heat_eq.ipynb

```diff
@@ -130,21 +134,536 @@
    "outputs": [],
    "source": [
     "local_domain =  mesh.create_rectangle(MPI.COMM_SELF,\n",
-    "                               [[0.,0.],[2.,3.]],\n",
-    "                               [15, 20], \n",
+    "                               [[0.,0.],[length, width]],\n",
+    "                               [Nx, Ny], \n",
     "                               cell_type=mesh.CellType.quadrilateral)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "7f792ea5-ce66-4f91-a2d7-10f5a361b9d6",
+   "metadata": {
+    "lines_to_next_cell": 2,
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Setting up a variational problem"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eea38f11-101f-4fd8-a8bb-af43979cc839",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "notes"
+    },
+    "tags": []
+   },
+   "source": [
+    "We will solve the heat equation, with a backward Euler time stepping scheme, i.e."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3578231c-f2ff-4cbe-beba-aeef1e7a3c40",
+   "metadata": {},
+   "source": [
+    "$$\n",
+    "\\begin{align*}\n",
+    "\\frac{u_{n+1}-u_n}{\\Delta t} - \\nabla \\cdot (\\mu  \\nabla u_{n+1}) &= f(x,t) \\qquad \\text{in } \\Omega,\\\\\n",
+    "u &= u_D(x,t) \\qquad \\text{on } \\partial\\Omega_D,\\\\\n",
+    "\\mu\\frac{\\partial u}{\\partial n} &=0 \\qquad \\text{on } \\partial\\Omega_N\n",
+    "\\end{align*}\n",
+    "$$ \n",
+    "with $u_D = \\sin(t)\\cos(y)$, $f=0$."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7fdc0d5f-e35b-4146-9839-ddf3a791f3c0",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "notes"
+    },
+    "tags": []
+   },
+   "source": [
+    "We start by defining the function space, the corresponding test and trial functions, as well as material and temporal parameters. We note that we use explicit imports from ufl to create the test and trial functions, to avoid confusion as to where they originate from. DOLFINx and UFL supports both real and complex valued functions. However, to be able to the PETSc linear algebra backend, which only supports a single floating type at compilation, we need to use appropriate scalar types in our variational form. This ensures that we generate consistent matrices/vectors"
```

Comment:
```suggestion
    "We start by defining the function space, the corresponding test and trial functions, as well as material and temporal parameters. We note that we use explicit imports from ufl to create the test and trial functions, to avoid confusion as to where they originate from. DOLFINx and UFL supports both real and complex valued functions. However, to be able to use the PETSc linear algebra backend, which only supports a single floating type at compilation, we need to use appropriate scalar types in our variational form. This ensures that we generate consistent matrices/vectors"
```